File: qqbot-admin-api/src/main/java/com/mieai/qqbot/admin/database/DatabaseConfigurationController.java
package com.mieai.qqbot.admin.database;

import jakarta.validation.Valid;
import java.util.Objects;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequestMapping("/api/system/database")
public class DatabaseConfigurationController {
    private final DatabaseAdministrationService service;

    public DatabaseConfigurationController(DatabaseAdministrationService service) {
        this.service = Objects.requireNonNull(service, "service must not be null");
    }

    @GetMapping("/configuration")
    public ResponseEntity<DatabaseConfigurationView> configuration() {
        return configurationResponse(service.current());
    }

    @PostMapping("/test")
    public DatabaseTestResult test(@Valid @RequestBody DatabaseCandidateRequest request) {
        DatabasePassword password = request.password() == null ? null : DatabasePassword.of(request.password());
        try (password) {
            return service.test(request.toSettings(password));
        }
    }

    @PostMapping("/switch")
    public ResponseEntity<DatabaseSwitchResult> switchDatabase(
            @Valid @RequestBody DatabaseSwitchRequest request,
            Authentication authentication) {
        DatabaseCandidateRequest candidate = request.candidate();
        DatabasePassword password = candidate.password() == null ? null : DatabasePassword.of(candidate.password());
        DatabaseSwitchResult result;
        try (password) {
            result = service.switchDatabase(
                    request.expectedRevision(),
                    candidate.toSettings(password),
                    authentication.getName());
        }
        return switchResponse(result);
    }

    @PostMapping("/reload")
    public ResponseEntity<DatabaseSwitchResult> reload(
            @Valid @RequestBody ReloadDatabaseRequest request,
            Authentication authentication) {
        return switchResponse(service.reload(request.expectedRevision(), authentication.getName()));
    }

    private static ResponseEntity<DatabaseConfigurationView> configurationResponse(
            DatabaseConfigurationView configuration) {
        return ResponseEntity.ok()
                .eTag(etag(configuration.revision()))
                .body(configuration);
    }

    private static ResponseEntity<DatabaseSwitchResult> switchResponse(DatabaseSwitchResult result) {
        return ResponseEntity.ok()
                .eTag(etag(result.configuration().revision()))
                .body(result);
    }

    private static String etag(long revision) {
        return '"' + Long.toString(revision) + '"';
    }
}

File: qqbot-module-database-support/src/main/kotlin/com/mieai/qqbot/admin/database/DatabaseConfigurationController.kt
package com.mieai.qqbot.admin.database

import jakarta.validation.Valid
import org.springframework.http.ResponseEntity
import org.springframework.security.core.Authentication
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/api/system/database")
class DatabaseConfigurationController(
    private val service: DatabaseAdministrationService,
) {
    @GetMapping("/configuration")
    fun configuration(): ResponseEntity<DatabaseConfigurationView> = configurationResponse(service.current())

    @PostMapping("/test")
    fun test(@Valid @RequestBody request: DatabaseCandidateRequest): DatabaseTestResult {
        val password = request.password?.let(DatabasePassword::of)
        try {
            return service.test(request.toSettings(password))
        } finally {
            password?.close()
        }
    }

    @PostMapping("/switch")
    fun switchDatabase(
        @Valid @RequestBody request: DatabaseSwitchRequest,
        authentication: Authentication,
    ): ResponseEntity<DatabaseSwitchResult> {
        val candidate = requireNotNull(request.candidate)
        val password = candidate.password?.let(DatabasePassword::of)
        try {
            val result = service.switchDatabase(
                request.expectedRevision,
                candidate.toSettings(password),
                authentication.name,
            )
            return switchResponse(result)
        } finally {
            password?.close()
        }
    }

    @PostMapping("/reload")
    fun reload(
        @Valid @RequestBody request: ReloadDatabaseRequest,
        authentication: Authentication,
    ): ResponseEntity<DatabaseSwitchResult> =
        switchResponse(service.reload(request.expectedRevision, authentication.name))

    private companion object {
        fun configurationResponse(configuration: DatabaseConfigurationView): ResponseEntity<DatabaseConfigurationView> =
            ResponseEntity.ok().eTag(etag(configuration.revision)).body(configuration)

        fun switchResponse(result: DatabaseSwitchResult): ResponseEntity<DatabaseSwitchResult> =
            ResponseEntity.ok().eTag(etag(result.configuration.revision)).body(result)

        fun etag(revision: Long): String = "\"$revision\""
    }
}
